internal/record: sync: unlock of unlocked mutex

Seen in a CI failure: https://travis-ci.org/cockroachdb/pebble/jobs/596159193

```
goroutine 10914 [running]:
runtime.throw(0x8836fd, 0x1e)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/runtime/panic.go:774 +0x72 fp=0xc004081528 sp=0xc0040814f8 pc=0x4304d2
sync.throw(0x8836fd, 0x1e)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/runtime/panic.go:760 +0x35 fp=0xc004081548 sp=0xc004081528 pc=0x430455
sync.(*Mutex).unlockSlow(0xc001486070, 0xffffffff)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/sync/mutex.go:196 +0xd6 fp=0xc004081570 sp=0xc004081548 pc=0x46bca6
sync.(*Mutex).Unlock(0xc001486070)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/sync/mutex.go:190 +0x48 fp=0xc004081590 sp=0xc004081570 pc=0x46bbb8
runtime.call32(0x0, 0x8973f0, 0xc004081798, 0x800000008)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/runtime/asm_amd64.s:539 +0x3b fp=0xc0040815c0 sp=0xc004081590 pc=0x45d5cb
panic(0x7da7a0, 0x912d60)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/runtime/panic.go:679 +0x1b2 fp=0xc004081650 sp=0xc0040815c0 pc=0x430012
sync.(*WaitGroup).Add(0xc003c526d0, 0xffffffffffffffff)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/sync/waitgroup.go:74 +0x139 fp=0xc004081688 sp=0xc004081650 pc=0x46d1f9
sync.(*WaitGroup).Done(...)
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/sync/waitgroup.go:99
github.com/cockroachdb/pebble/internal/record.(*syncQueue).pop(0xc0014860f0, 0x2, 0x0, 0x0, 0x1c, 0x0)
	/home/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:128 +0x5f fp=0xc0040816e8 sp=0xc004081688 pc=0x54565f
github.com/cockroachdb/pebble/internal/record.(*LogWriter).flushLoop(0xc001486000)
	/home/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:312 +0x33a fp=0xc0040817d8 sp=0xc0040816e8 pc=0x545f9a
runtime.goexit()
	/home/travis/.gimme/versions/go1.13.1.linux.amd64/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0040817e0 sp=0xc0040817d8 pc=0x45f2e1
created by github.com/cockroachdb/pebble/internal/record.NewLogWriter
	/home/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:265 +0x2c4
```

I believe the real problem is that a panic was thrown in the `flushLoop`. 

The only stack trace which shows a test running shows `TestLogData`. I tried stressing that test locally and couldn't reproduce a failure. The failure happened on go1.13, so perhaps that is an important factor.

Saw this again on CI, and now I know how to reproduce:

```
~ make stress PKG=. TESTS='TestLargeBatch|TestLogData'
GO111MODULE=off go test -v  -exec 'stress ' -run "TestLargeBatch|TestLogData" -timeout 0 .

=== RUN   TestLargeBatch


0:
  5:[a-a]
--- PASS: TestLargeBatch (0.00s)
=== RUN   TestLogData
--- FAIL: TestLogData (0.00s)
    db_test.go:529: <nil>
        goroutine 39 [running]:
        runtime/debug.Stack(0xc00022bd10, 0x13e1760, 0x151bd30)
        	/Users/pmattis/Development/go1.13/src/runtime/debug/stack.go:24 +0x9d
        github.com/cockroachdb/pebble/internal/record.(*LogWriter).flushPending.func1(0xc000067f30)
        	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:329 +0x46
        panic(0x13e1760, 0x151bd30)
        	/Users/pmattis/Development/go1.13/src/runtime/panic.go:679 +0x1b2
        sync.(*WaitGroup).Add(0xc0001c20b0, 0xffffffffffffffff)
        	/Users/pmattis/Development/go1.13/src/sync/waitgroup.go:74 +0x139
        sync.(*WaitGroup).Done(...)
        	/Users/pmattis/Development/go1.13/src/sync/waitgroup.go:99
        github.com/cockroachdb/pebble/internal/record.(*syncQueue).pop(0xc0001f65e8, 0x1, 0x0, 0x0, 0x1c, 0x0)
        	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:129 +0x5f
        github.com/cockroachdb/pebble/internal/record.(*LogWriter).flushPending(0xc0001f6500, 0xc00068e008, 0x1c, 0x8000, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0)
        	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:346 +0x12b
        github.com/cockroachdb/pebble/internal/record.(*LogWriter).flushLoop(0xc0001f6500)
        	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:309 +0x1ed
        created by github.com/cockroachdb/pebble/internal/record.NewLogWriter
        	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:265 +0x28e

FAIL
```

I also know what the problem is. Fix to arrive shortly.

--- a/batch.go
+++ b/batch.go
@@ -47,6 +47,11 @@ type batchStorage struct {
 	//     - the varint-string user key,
 	//     - the varint-string value (if kind != delete).
 	// The sequence number and count are stored in little-endian order.
+	//
+	// The data field can be (but is not guaranteed to be) nil for new
+	// batches. Large batches will set the data field to nil when committed as
+	// the data has been moved to a flushableBatch and inserted into the queue of
+	// memtables.
 	data           []byte
 	cmp            Compare
 	abbreviatedKey AbbreviatedKey
@@ -277,9 +282,20 @@ func newIndexedBatch(db *DB, comparer *Comparer) *Batch {
 }
 
 func (b *Batch) release() {
-	// NB: This is ugly, but necessary so that we can use atomic.StoreUint32 for
-	// the Batch.applied field. Without using an atomic to clear that field the
-	// Go race detector complains.
+	if b.db == nil {
+		// The batch was not created using newBatch or newIndexedBatch, or an error
+		// was encountered. We don't try to reuse batches that encountered an error
+		// because they might be stuck somewhere in the system and attempting to
+		// reuse such batches is a recipe for onerous debugging sessions. Instead,
+		// let the GC do its job.
+		return
+	}
+	b.db = nil
+
+	// NB: This is ugly (it would be cleaner if we could just assign a Batch{}),
+	// but necessary so that we can use atomic.StoreUint32 for the Batch.applied
+	// field. Without using an atomic to clear that field the Go race detector
+	// complains.
 	b.Reset()
 	b.storage.cmp = nil
 	b.storage.abbreviatedKey = nil
@@ -289,13 +305,6 @@ func (b *Batch) release() {
 	b.commit = sync.WaitGroup{}
 	b.commitErr = nil
 	atomic.StoreUint32(&b.applied, 0)
-
-	if b.db == nil {
-		// Batch not created using newBatch or newIndexedBatch, so don't put it
-		// back in the pool.
-		return
-	}
-	b.db = nil
 
 	if b.index == nil {
 		batchPool.Put(b)
@@ -759,6 +768,7 @@ func (b *Batch) init(cap int) {
 // where the end result is a Repr() call, not a Commit call or a Close call.
 // Commits and Closes take care of releasing resources when appropriate.
 func (b *Batch) Reset() {
+	b.count = 0
 	if b.storage.data != nil {
 		if cap(b.storage.data) > batchMaxRetainedSize {
 			// If the capacity of the buffer is larger than our maximum
@@ -771,7 +781,6 @@ func (b *Batch) Reset() {
 			b.storage.data = b.storage.data[:batchHeaderLen]
 			b.setSeqNum(0)
 		}
-		b.count = 0
 	}
 }
 

--- a/commit.go
+++ b/commit.go
@@ -240,18 +240,19 @@ func (p *commitPipeline) Commit(b *Batch, syncWAL bool) error {
 	// Prepare the batch for committing: enqueuing the batch in the pending
 	// queue, determining the batch sequence number and writing the data to the
 	// WAL.
+	//
+	// NB: We set Batch.commitErr on error so that the batch won't be a candidate
+	// for reuse. See Batch.release().
 	mem, err := p.prepare(b, syncWAL)
 	if err != nil {
-		// TODO(peter): what to do on error? the pipeline will be horked at this
-		// point.
-		panic(err)
+		b.db = nil // prevent batch reuse on error
+		return err
 	}
 
 	// Apply the batch to the memtable.
 	if err := p.env.apply(b, mem); err != nil {
-		// TODO(peter): what to do on error? the pipeline will be horked at this
-		// point.
-		panic(err)
+		b.db = nil // prevent batch reuse on error
+		return err
 	}
 
 	// Publish the batch sequence number.
@@ -260,11 +261,9 @@ func (p *commitPipeline) Commit(b *Batch, syncWAL bool) error {
 	<-p.sem
 
 	if b.commitErr != nil {
-		// TODO(peter): what to do on a log sync error? the pipeline will be horked
-		// at this point.
-		panic(b.commitErr)
+		b.db = nil // prevent batch reuse on error
 	}
-	return nil
+	return b.commitErr
 }
 
 // AllocateSeqNum allocates count sequence numbers, invokes the prepare

--- a/commit.go
+++ b/commit.go
@@ -227,9 +227,6 @@ func newCommitPipeline(env commitEnv) *commitPipeline {
 	return p
 }
 
-func (p *commitPipeline) Close() {
-}
-
 // Commit the specified batch, writing it to the WAL, optionally syncing the
 // WAL, and applying the batch to the memtable. Upon successful return the
 // batch's mutations will be visible for reading.

--- a/db.go
+++ b/db.go
@@ -338,9 +338,13 @@ func (d *DB) getInternal(key []byte, b *Batch, s *Snapshot) ([]byte, error) {
 // It is safe to modify the contents of the arguments after Set returns.
 func (d *DB) Set(key, value []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.Set(key, value, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // Delete deletes the value for the given key. Deletes are blind all will
@@ -349,9 +353,13 @@ func (d *DB) Set(key, value []byte, opts *WriteOptions) error {
 // It is safe to modify the contents of the arguments after Delete returns.
 func (d *DB) Delete(key []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.Delete(key, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // SingleDelete adds an action to the batch that single deletes the entry for key.
@@ -360,9 +368,13 @@ func (d *DB) Delete(key []byte, opts *WriteOptions) error {
 // It is safe to modify the contents of the arguments after SingleDelete returns.
 func (d *DB) SingleDelete(key []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.SingleDelete(key, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // DeleteRange deletes all of the keys (and values) in the range [start,end)
@@ -372,9 +384,13 @@ func (d *DB) SingleDelete(key []byte, opts *WriteOptions) error {
 // returns.
 func (d *DB) DeleteRange(start, end []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.DeleteRange(start, end, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // Merge adds an action to the DB that merges the value at key with the new
@@ -384,9 +400,13 @@ func (d *DB) DeleteRange(start, end []byte, opts *WriteOptions) error {
 // It is safe to modify the contents of the arguments after Merge returns.
 func (d *DB) Merge(key, value []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.Merge(key, value, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // LogData adds the specified to the batch. The data will be written to the
@@ -396,9 +416,13 @@ func (d *DB) Merge(key, value []byte, opts *WriteOptions) error {
 // It is safe to modify the contents of the argument after LogData returns.
 func (d *DB) LogData(data []byte, opts *WriteOptions) error {
 	b := newBatch(d)
-	defer b.release()
 	_ = b.LogData(data, opts)
-	return d.Apply(b, opts)
+	if err := d.Apply(b, opts); err != nil {
+		return err
+	}
+	// Only release the batch on success.
+	b.release()
+	return nil
 }
 
 // Apply the operations contained in the batch to the DB. If the batch is large
@@ -429,20 +453,22 @@ func (d *DB) Apply(batch *Batch, opts *WriteOptions) error {
 	if int(batch.memTableSize) >= d.largeBatchThreshold {
 		batch.flushable = newFlushableBatch(batch, d.opts.Comparer)
 	}
-	err := d.commit.Commit(batch, sync)
-	if err == nil {
-		// If this is a large batch, we need to clear the batch contents as the
-		// flushable batch may still be present in the flushables queue.
-		//
-		// TODO(peter): Currently large batches are written to the WAL. We could
-		// skip the WAL write and instead wait for the large batch to be flushed to
-		// an sstable. For a 100 MB batch, this might actually be faster. For a 1
-		// GB batch this is almost certainly faster.
-		if batch.flushable != nil {
-			batch.storage.data = nil
-		}
+	if err := d.commit.Commit(batch, sync); err != nil {
+		// There isn't much we can do on an error here. The commit pipeline will be
+		// horked at this point.
+		d.opts.Logger.Fatalf("%v", err)
 	}
-	return err
+	// If this is a large batch, we need to clear the batch contents as the
+	// flushable batch may still be present in the flushables queue.
+	//
+	// TODO(peter): Currently large batches are written to the WAL. We could
+	// skip the WAL write and instead wait for the large batch to be flushed to
+	// an sstable. For a 100 MB batch, this might actually be faster. For a 1
+	// GB batch this is almost certainly faster.
+	if batch.flushable != nil {
+		batch.storage.data = nil
+	}
+	return nil
 }
 
 func (d *DB) commitApply(b *Batch, mem *memTable) error {

--- a/db.go
+++ b/db.go
@@ -723,7 +723,6 @@ func (d *DB) Close() error {
 		panic("pebble: log-writer should be nil in read-only mode")
 	}
 	err = firstError(err, d.fileLock.Close())
-	d.commit.Close()
 
 	err = firstError(err, d.dataDir.Close())
 	if d.dataDir != d.walDir {

--- a/db_test.go
+++ b/db_test.go
@@ -426,6 +426,13 @@ func TestLargeBatch(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	// Allocate a bunch of batches to exhaust the batchPool. None of these
+	// batches should have a non-zero count.
+	for i := 0; i < 10; i++ {
+		b := d.NewBatch()
+		require.EqualValues(t, 0, b.Count())
+	}
+
 	if err := d.Close(); err != nil {
 		t.Fatal(err)
 	}

--- a/mem_table.go
+++ b/mem_table.go
@@ -225,7 +225,7 @@ func (m *memTable) apply(batch *Batch, seqNum uint64) error {
 		}
 	}
 	if seqNum != startSeqNum+uint64(batch.Count()) {
-		panic(fmt.Sprintf("pebble: inconsistent batch count: %d vs %d",
+		panic(fmt.Errorf("pebble: inconsistent batch count: %d vs %d",
 			seqNum, startSeqNum+uint64(batch.Count())))
 	}
 	if tombstoneCount != 0 {

